File: frontend-nextjs/src/core/utils/sanitizer.ts
/**
 * Gurukul AI Student-Safe Content Sanitizer
 *
 * Deterministically removes extraction artifacts while preserving
 * legitimate educational prose.
 */

const ARTIFACT_PATTERNS = [
  /Chapter\s+\d+\.indd\s+\d+\s+\d{2}-\d{2}-\d{4}\s+\d{2}:\d{2}:\d{2}/gi, // InDesign markers
  /\d{2}-\d{2}-\d{4}\s+\d{2}:\d{2}:\d{2}/g, // Timestamps
  /Grade\s+\d+/gi, // Grade markers
  /Santoor\s+Grade\s+\d+/gi, // Specific book markers
  /\[Concept:\s*Concept\s*\d+\]/gi, // Internal concept bracket tags
];

export function sanitizeStudentContent(text: string): string {
  if (!text) return "";

  let cleaned = text;

  // 1. Apply regex patterns for known artifacts
  ARTIFACT_PATTERNS.forEach(pattern => {
    cleaned = cleaned.replace(pattern, "");
  });

  // 2. Clean up resulting messy whitespace
  cleaned = cleaned.replace(/\s{2,}/g, " ").trim();

  // 3. Handle specific labels like UNSPECIFIED
  if (cleaned.toUpperCase() === "UNSPECIFIED") {
    return "Not specified";
  }

  return cleaned;
}

/**
 * Specifically cleans question stems which often have repetitive metadata framing.
 */
export function sanitizeQuestionStem(stem: string, conceptName?: string): string {
  let cleaned = sanitizeStudentContent(stem);

  // If the stem is a template like "Explain the important information in about...",
  // we try to make it more direct.
  if (conceptName) {
     cleaned = cleaned.replace(new RegExp(`about '${conceptName}'`, 'gi'), "");
     cleaned = cleaned.replace(new RegExp(`about ${conceptName}`, 'gi'), "");
  }

  return cleaned;
}

File: backend/src/core/utils/sanitizer.py
import re

ARTIFACT_PATTERNS = [
    r"Chapter\s+\d+\.indd\s+\d+\s+\d{2}-\d{2}-\d{4}\s+\d{2}:\d{2}:\d{2}", # Full InDesign markers
    r"Chapter\s+\d+\.indd\s+\d+", # Short InDesign markers
    r"Chapter\s+\d+\.indd",
    r"\w+\.indd\s+\d+", # Any indd and numbers
    r"\d+\.indd\s+\d+", # Just indd and numbers
    r"\d{2,}-\d{2}-\d{4}\s+\d{2}:\d{2}:\d{2}", # Timestamps with long day/month
    r"\d{2}-\d{2}-\d{4}\s+\d{2}:\d{2}:\d{2}", # Standard Timestamps
    r"\d{2,}-\d{2}-\d{4}", # Date sequences
    r"\d{2}:\d{2}:\d{2}", # Time sequences
    r"2\s+3\s+5\s+6\s+4\s+1\s+7\s+8\s+9", # Specific extraction sequence
    r"\d\s+\d\s+\d\s+\d\s+\d\s+\d\s+\d\s+\d\s+\d", # Generic long number sequences
    r"Grade\s+\d+",
    r"Santoor\s+Grade\s+\d+",
    r"Unit\s+\d+:\s+Let’s\s+Have\s+Fun",
    r"Let us Recite \d+",
    r"Canonical source evidence",
    r"source-supported points below",
    r"source-supported statement",
    r"supplied source evidence",
    r"important information in evidence",
    r"evidence \d+",
    r"Concept\s+\d+", # Remove generic 'Concept X' text from strings
]

def sanitize_student_content(text: str) -> str:
    if not text:
        return ""

    cleaned = text
    for pattern in ARTIFACT_PATTERNS:
        cleaned = re.sub(pattern, "", cleaned, flags=re.IGNORECASE)

    # Clean up messy whitespace
    cleaned = re.sub(r"\s{2,}", " ", cleaned).strip()

    if cleaned.upper() == "UNSPECIFIED":
        return "Not specified"

    return cleaned

def sanitize_question_stem(stem: str, concept_name: str = None) -> str:
    cleaned = sanitize_student_content(stem)
    if concept_name:
        cleaned = cleaned.replace(f"about '{concept_name}'", "")
        cleaned = cleaned.replace(f"about {concept_name}", "")
    return cleaned
